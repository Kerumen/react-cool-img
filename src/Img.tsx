import React from 'react';

export default () => (
  <div>
    <h1>Hello Img!</h1>
  </div>
);
