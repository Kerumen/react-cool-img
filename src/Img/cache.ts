const key = 'REACT_COOL_IMG';

export const set = (src: string): void => {
  try {
    const data = JSON.parse(sessionStorage.getItem(key)) || {};
    data[src] = new Date();
    sessionStorage.setItem(key, JSON.stringify(data));
  } catch (error) {
    console.error(`ðŸ’¡react-cool-img: ${error}`);
  }
};

export const get = (src: string): boolean => {
  try {
    const data = JSON.parse(sessionStorage.getItem(key)) || {};
    return !!data[src] || false;
  } catch (error) {
    console.error(`ðŸ’¡react-cool-img: ${error}`);
    return false;
  }
};
