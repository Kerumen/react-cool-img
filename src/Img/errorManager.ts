const prefix = 'ðŸ¤¡react-cool-img:';
export const msgs = {
  decode: (src: string): string => `${prefix} error decoding image at ${src}`,
  onerror: (src: string): string => `${prefix} error loading image at ${src}`,
  retry: `${prefix} to use retry, you must setup "count" and "delay"`
};

interface Params {
  src?: string;
  retry?: { count?: number; delay?: number };
}

export default (type: string, { src, retry }: Params): void => {
  switch (type) {
    case 'decode':
      console.error(msgs[type](src));
      break;
    case 'onerror':
      console.error(msgs[type](src));
      break;
    case 'retry':
      if (retry && !(retry.count && retry.delay)) console.error(msgs[type]);
      break;
    default:
  }
};
