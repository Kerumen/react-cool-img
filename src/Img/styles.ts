export default (url: string, css: any) => css`
  background: url(${url}) no-repeat center;
  background-size: 100% 100%;
`;
