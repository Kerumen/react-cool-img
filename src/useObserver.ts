import { Dispatch, useState, useRef, useEffect, useCallback } from 'react';

export const observerErr =
  "ðŸ’¡react-cool-img: the browser doesn't support Intersection Observer, please install polyfill to enable lazy loading: https://github.com/wellyshen/react-cool-img#intersection-observer-polyfill";
export const thresholdErr =
  'ðŸ’¡react-cool-img: the threshold of observerOptions must be a number. Use 0 as fallback.';

export interface Options {
  root?: HTMLElement;
  rootMargin?: string;
  threshold?: number;
}
export type Return = readonly [
  Dispatch<HTMLElement | null>,
  boolean,
  Dispatch<boolean>
];

export default (
  debounce: number,
  { root, rootMargin = '50px', threshold = 0.01 }: Options
): Return => {
  const [startLoad, setStartLoad] = useState<boolean>(false);
  const [el, setEl] = useState<HTMLElement | null>(null);
  const observerRef = useRef<IntersectionObserver>(null);
  const timeoutRef = useRef<NodeJS.Timeout>(null);
  let numThreshold = threshold;

  if (typeof threshold !== 'number') {
    console.error(thresholdErr);
    numThreshold = 0;
  }

  const resetTimeout = useCallback((): void => {
    if (!timeoutRef.current) return;

    clearTimeout(timeoutRef.current);
    timeoutRef.current = null;
  }, []);

  useEffect(() => {
    if (
      !('IntersectionObserver' in window) ||
      !('IntersectionObserverEntry' in window) ||
      !('isIntersecting' in window.IntersectionObserverEntry.prototype)
    ) {
      console.error(observerErr);
      return (): void => null;
    }

    observerRef.current = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting && !startLoad) {
          timeoutRef.current = setTimeout(() => {
            setStartLoad(true);
          }, debounce);
        } else {
          resetTimeout();
        }
      },
      { root, rootMargin, threshold: numThreshold }
    );

    const { current: observer } = observerRef;

    if (el) observer.observe(el);

    return (): void => {
      observer.disconnect();
      resetTimeout();
    };
  }, [el, startLoad, root, rootMargin, numThreshold, debounce, resetTimeout]);

  // setStartLoad is used for testing
  return [setEl, startLoad, setStartLoad];
};
