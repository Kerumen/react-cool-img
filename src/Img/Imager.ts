/* eslint-disable lines-between-class-members */

import errorManager from './errorManager';

export interface Retry {
  count: number;
  delay: number;
  acc?: '+' | '*' | boolean;
}

export default class Imager {
  img: HTMLImageElement | null;
  retries: number;
  timeout: any;

  constructor() {
    this.img = null;
    this.timeout = null;
    this.retries = 1;
  }

  load(
    src: string,
    crossOrigin: string | null,
    decode: boolean,
    retry: Retry | null,
    onError: (event: Event) => void,
    onLoad: (event: Event) => void
  ): void {
    this.img = new Image();
    this.img.src = src;

    if (crossOrigin) this.img.crossOrigin = crossOrigin;
    if (decode)
      this.img.decode().catch(() => {
        errorManager('decode', { src });
      });

    this.img.onerror = (event: Event): void => {
      errorManager('retry', { retry });

      if (
        !retry ||
        !retry.count ||
        !retry.delay ||
        this.retries > retry.count
      ) {
        onError(event);
        return;
      }

      let time = retry.delay ** this.retries;
      if (retry.acc === '+') time = retry.delay * this.retries;
      if (retry.acc === false) time = retry.delay;

      this.timeout = setTimeout(() => {
        this.clearImgSrc();
        this.img.src = src;
      }, time * 1000);

      this.retries += 1;
    };
    this.img.onload = (event: Event): void => {
      onLoad(event);
    };
  }

  unload(): void {
    if (this.img) {
      this.img.onerror = null;
      this.img.onload = null;
      this.clearImgSrc();
      this.img = null;
    }
    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = null;
    }
    this.retries = 1;
  }

  clearImgSrc(): void {
    this.img.src = '';
    try {
      delete this.img.src;
    } catch (error) {
      // Ignore the error of deleting object properties in Safari strict mode
    }
  }
}
