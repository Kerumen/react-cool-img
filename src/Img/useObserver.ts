import { useState, useRef, useEffect } from 'react';

import errorManager from './errorManager';

export interface Config {
  root?: Element;
  rootMargin?: string;
  threshold?: number;
  debounce?: number;
}
export type Return = [
  (node?: Element | null) => void,
  boolean,
  (val: boolean) => void
];

export default (
  lazy: boolean,
  { root = null, rootMargin = '50px', threshold = 0.01, debounce = 300 }: Config
): Return => {
  if (!lazy || typeof window === 'undefined' || !window.IntersectionObserver) {
    if (!window.IntersectionObserver) errorManager('observer');

    const setState = (): void => {};
    return [setState, true, setState];
  }

  const [startLoad, setStartLoad] = useState(false);
  const [node, setNode] = useState(null);
  const observerRef = useRef(null);
  const timeoutRef = useRef(null);
  let numThreshold = threshold;

  if (typeof threshold !== 'number') {
    errorManager('threshold');
    numThreshold = 0;
  }

  const resetTimeout = (): void => {
    if (!timeoutRef.current) return;

    clearTimeout(timeoutRef.current);
    timeoutRef.current = null;
  };

  useEffect(() => {
    if (observerRef.current) observerRef.current.disconnect();

    observerRef.current = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting && !startLoad) {
          timeoutRef.current = setTimeout(() => {
            setStartLoad(true);
          }, debounce);
        } else {
          resetTimeout();
        }
      },
      { root, rootMargin, threshold: numThreshold }
    );

    const { current: observer } = observerRef;

    if (node) observer.observe(node);

    return (): void => {
      observer.disconnect();
      resetTimeout();
    };
  }, [node, startLoad, root, rootMargin, numThreshold, debounce]);

  // setStartLoad is used for testing
  return [setNode, startLoad, setStartLoad];
};
