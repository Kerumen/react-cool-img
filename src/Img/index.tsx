import React, {
  SFC,
  SyntheticEvent,
  ReactNode,
  useState,
  useEffect,
  memo
} from 'react';
import { ClassNames } from '@emotion/core';

import loadImgSize from '../utils/loadImgSize';
import styles from './styles';

interface Props {
  src?: string;
  placeholderSrc?: string;
  placeholderAsError?: boolean;
  errorSrc?: string;
  width?: string;
  height?: string;
  alt?: string;
  className?: string;
  onLoad?: (event: SyntheticEvent) => void;
  onError?: (event: SyntheticEvent) => void;
}

const Img: SFC<Props> = ({
  src,
  placeholderSrc,
  placeholderAsError,
  errorSrc,
  width,
  height,
  alt,
  className,
  onLoad,
  onError
}: Props) => {
  const [source, setSource] = useState(src);
  const [imgSize, setImgSize] = useState({ width: null, height: null });
  const [showPlaceholder, setShowPlaceholder] = useState(!!placeholderSrc);

  useEffect(() => {
    if (placeholderSrc)
      loadImgSize(placeholderSrc, width, height, (w, h) => {
        setImgSize({ width: w, height: h });
      });
  }, [placeholderSrc, width, height]);

  const handleLoaded = (event: SyntheticEvent): void => {
    onLoad(event);
    setShowPlaceholder(false);
  };

  const handleError = (event: SyntheticEvent): void => {
    onError(event);

    if (errorSrc) {
      setSource(errorSrc);
    } else if (placeholderAsError && placeholderSrc) {
      setSource(placeholderSrc);
    } else {
      setShowPlaceholder(false);
    }
  };

  return (
    <ClassNames>
      {({ cx, css }): ReactNode => (
        <img
          className={cx(
            { [styles(placeholderSrc, css)]: showPlaceholder },
            className
          )}
          src={source}
          width={width || showPlaceholder ? imgSize.width : null}
          height={height || showPlaceholder ? imgSize.height : null}
          alt={alt}
          onLoad={handleLoaded}
          onError={handleError}
        />
      )}
    </ClassNames>
  );
};

Img.defaultProps = {
  src: null,
  placeholderSrc: null,
  placeholderAsError: true,
  errorSrc: null,
  width: null,
  height: null,
  alt: null,
  className: '',
  onLoad: (): void => {},
  onError: (): void => {}
};

export default memo(Img);
