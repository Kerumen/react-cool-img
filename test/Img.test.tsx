/* eslint-disable import/first */

const FAILURE_SRC = 'FAILURE_SRC';
const SUCCESS_SRC = 'SUCCESS_SRC';

const setState = (): void => null;
jest.mock('../src/Img/useObserver', () =>
  jest.fn(() => [setState, false, setState])
);

const set = jest.fn();
const get = jest.fn(() => false);
jest.mock('../src/Img/storage', () => ({ set, get }));

const load = jest.fn((...args) =>
  args[args[0] === FAILURE_SRC ? 4 : 5]({ target: { src: args[0] } })
);
const unload = jest.fn();
jest.mock('../src/Img/Imager', () => jest.fn(() => ({ load, unload })));

import React, { ReactElement } from 'react';
import { render } from '@testing-library/react';

import useObserver from '../src/Img/useObserver';
import * as storage from '../src/Img/storage';
import Img from '../src/Img';

describe('<Img />', () => {
  const props = {
    className: 'cool-image',
    placeholder: 'PLACEHOLDER_SRC',
    error: 'ERROR_SRC',
    crossOrigin: 'anonymous' as const,
    decode: true,
    lazy: true,
    debounce: 300,
    observerOptions: { rootMargin: '50px', threshold: 0.01 },
    retry: { count: 5, delay: 2 },
    srcSet: 'cool.png',
    sizes: '100vw',
    onError: jest.fn(),
    onLoad: jest.fn(),
    alt: 'Cool Image'
  };

  const matchSnapshot = (img: ReactElement): void => {
    expect(render(img).asFragment()).toMatchSnapshot();
  };
  const setStartLoad = (val: boolean): void => {
    // @ts-ignore
    useObserver.mockImplementation(() => [setState, val, setState]);
  };

  beforeEach(() => {
    // @ts-ignore
    storage.get.mockReset();
  });

  it("should setup useObserver's arguments correctly", () => {
    render(<Img src={SUCCESS_SRC} {...props} />);

    const { lazy, debounce, observerOptions } = props;

    expect(useObserver).toBeCalledWith(lazy, debounce, observerOptions);
  });

  it('should unload src image', () => {
    const { unmount } = render(<Img src={SUCCESS_SRC} {...props} />);

    unmount();
    expect(unload).toBeCalled();
  });

  it('should render placeholder image', () => {
    matchSnapshot(<Img src={SUCCESS_SRC} {...props} />);

    expect(load).not.toBeCalled();
    expect(props.onLoad).not.toBeCalled();
  });

  it('should render default placeholder image', () => {
    matchSnapshot(<Img src={SUCCESS_SRC} {...props} placeholder={null} />);
  });

  it('should render placeholder image due to cache disabled', () => {
    // @ts-ignore
    storage.get.mockImplementation(() => true);

    matchSnapshot(<Img src={SUCCESS_SRC} {...props} cache={false} />);

    expect(set).not.toBeCalled();
  });

  it('should render src image', () => {
    setStartLoad(true);
    matchSnapshot(<Img src={SUCCESS_SRC} {...props} />);

    const { crossOrigin, decode, retry, onLoad } = props;

    expect(load).toBeCalledWith(
      SUCCESS_SRC,
      crossOrigin,
      decode,
      retry,
      expect.any(Function),
      expect.any(Function)
    );
    expect(onLoad).toBeCalled();
    expect(set).toBeCalledWith(SUCCESS_SRC);
  });

  it('should render src image due to image cached', () => {
    // @ts-ignore
    storage.get.mockImplementation(() => true);

    setStartLoad(false);
    matchSnapshot(<Img src={SUCCESS_SRC} {...props} />);

    const { crossOrigin, decode, retry, onLoad } = props;

    expect(load).toBeCalledWith(
      SUCCESS_SRC,
      crossOrigin,
      decode,
      retry,
      expect.any(Function),
      expect.any(Function)
    );
    expect(onLoad).toBeCalled();
    expect(set).toBeCalledWith(SUCCESS_SRC);
  });

  it('should render error image', () => {
    setStartLoad(true);
    matchSnapshot(<Img src={FAILURE_SRC} {...props} />);
  });

  it('should render placeholder image instead of error image', () => {
    setStartLoad(true);
    matchSnapshot(<Img src={FAILURE_SRC} {...props} error={null} />);
  });
});
