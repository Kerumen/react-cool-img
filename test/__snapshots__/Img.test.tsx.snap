// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Img /> should not render placeholder image 1`] = `
<DocumentFragment>
  <img
    alt="Cool"
    class=" no-js-SUCCESS_SRC"
    crossorigin="anonymous"
    sizes="100vw"
    src="SUCCESS_SRC"
    srcset="cool.png"
  />
  <noscript />
</DocumentFragment>
`;

exports[`<Img /> should render error image 1`] = `
<DocumentFragment>
  <img
    alt="Cool"
    class=" no-js-FAILURE_SRC"
    src="ERROR"
  />
  <noscript />
</DocumentFragment>
`;

exports[`<Img /> should render placeholder image 1`] = `
<DocumentFragment>
  <img
    alt="Cool"
    class=" no-js-SUCCESS_SRC"
    src="PLACEHOLDER"
  />
  <noscript />
</DocumentFragment>
`;

exports[`<Img /> should render placeholder image insteadly 1`] = `
<DocumentFragment>
  <img
    alt="Cool"
    class=" no-js-FAILURE_SRC"
    src="PLACEHOLDER"
  />
  <noscript />
</DocumentFragment>
`;

exports[`<Img /> should render src image 1`] = `
<DocumentFragment>
  <img
    alt="Cool"
    class=" no-js-SUCCESS_SRC"
    crossorigin="anonymous"
    sizes="100vw"
    src="SUCCESS_SRC"
    srcset="cool.png"
  />
  <noscript />
</DocumentFragment>
`;
